### PR TITLE
[WIP] Basic WM

### PR DESCRIPTION
**Pull request by charafau**
Done:

Remove circular dependency between compositor and surface
Add equals methods to Surface class
Change single surface view to multiple views in demo project
Commented out focusing code in surface for now

Todo:
Get actual surface size from wlroots
Actual floating and setting size to surface view
Add isPopUp variable to surface object

### CODE REVIEW

File: compositor_dart/lib/compositor_dart.dart [hunk 1]
```diff
@@ -111,17 +128,19 @@ class Compositor {
...
     platform.addHandler("surface_unmap", (call) async {
       int handle = call.arguments["handle"];
-      Surface surface = surfaces[handle]!;
-      surfaces.remove(handle);
-      surfaceUnmapped.add(surface);
+      if (surfaces.containsKey(handle)) {
+        Surface surface = surfaces[handle]!;
+        surfaces.remove(handle);
+        surfaceUnmapped.add(surface);
+      }
```
Comment (hansihe) [inline compositor_dart/lib/compositor_dart.dart:143]: For any given surface, I believe `map` and `unmap` should be only called strictly interspersed in a sequence:

`map`, `unmap`, `map`, `unmap`, ...

This should mean that:
* When `map` is called, there should **never** be a surface of that handle in the `surfaces` map, we add it
* When `unmap` is called, there should **always** be a surface of that handle in the `surfaces`, we remove it

We should likely add assertions for this, if this assumption is broken it would be a bug in the native code

File: compositor_dart/lib/compositor_dart.dart [hunk 2]
```diff
@@ -21,15 +22,33 @@ class Surface {
...
   final int uid;
 
-  final Compositor compositor;
-
   Surface({
     required this.handle,
     required this.pid,
     required this.gid,
     required this.uid,
-    required this.compositor,
```
Comment (hansihe) [inline compositor_dart/lib/compositor_dart.dart:31]: Should we keep a reference to `Compositor` from `Surface`s or not?

File: src/surface.c [hunk 3]
```diff
@@ -27,17 +27,18 @@ static void focus_view(struct fwr_view *view, struct wlr_surface *surface) {
...
+	// 	/*
+	// 	 * Deactivate the previously focused surface. This lets the client know
+	// 	 * it no longer has focus and the client will repaint accordingly, e.g.
+	// 	 * stop displaying a caret.
+	// 	 */
+	// 	struct wlr_xdg_surface *previous = wlr_xdg_surface_from_wlr_surface(
+	// 				seat->keyboard_state.focused_surface);
+	// 	assert(previous->role == WLR_XDG_SURFACE_ROLE_TOPLEVEL);
+	// 	wlr_xdg_toplevel_set_activated(previous->toplevel, false);
+	// }
```
Comment (hansihe) [inline src/surface.c:41]: Focus should definitely be managed by the dart side, but does setting more than one surface as focused violate a wlroots invariant?

In which case it seems like a good idea to keep guard code that makes sure the previous surface is unfocused when we attempt to set a new focus.

File: compositor_dart/lib/compositor_dart.dart (hunk 2 above)
Comment (charafau) [reply compositor_dart/lib/compositor_dart.dart:31]: I think it's not necassry. Compositor will be created in main, and probably main should keep track of it. Is there actually any use case where we might have more than 1 compositor ?

File: src/surface.c (hunk 3 above)
Comment (charafau) [reply src/surface.c:41]: I get exception from wlroots when I try to show second view and this `if` statement is here. Needs some fixes when we have proper focus management

File: compositor_dart/lib/compositor_dart.dart (hunk 2 above)
Comment (hansihe) [reply compositor_dart/lib/compositor_dart.dart:31]: There will never be more than one compositor at a time in a given isolate, but the `Compositor` it belongs to is still part of the state conceptually in my mind.

At the very least it prevents us from having to pass it around everywhere along with the surface. Most times you have a `Surface` you need access to the compositor as well.

In my mind we have a few options:
1. Pass the compositor object around alongside every surface. Might result in quite a bit of boilerplate as the dart side library expands in size.
2. Just rely on the fact that there will only ever be one compositor object. Access it through a global everywhere. Not sure I like this, I like to keep state local even with singletons.
3. Store the `Compositor` object within each surface. Only downside I can think of here is memory, but that's only a pointer.

File: compositor_dart/lib/compositor_dart.dart (hunk 1 above)
Comment (hansihe) [reply compositor_dart/lib/compositor_dart.dart:147]: I wonder if the fact that things fail here on multiple surfaces means there is an error somewhere in my code where multiple surfaces get the same handle id? I think there are some debug prints for when a surface is mapped with its handle id, are those different?

Comment (charafau) [reply compositor_dart/lib/compositor_dart.dart:147]: From what I've checked, handles were different but I think it was the way we managed compositor - now it's created in and managed in main.

File: src/surface.c (hunk 3 above)
Comment (charafau) [reply src/surface.c:42]: I will check today and let you know

File: compositor_dart/lib/compositor_dart.dart (hunk 2 above)
Comment (HrX03) [reply compositor_dart/lib/compositor_dart.dart:31]: what about providing the compositor using an `InheritedWidget`?
the app in order to use the compositor will need to use as root a `Compositor` widget of some sort which will then be used to create, maintain and dispose the compositor instance and also expose it to descendants using `Compositor.of(context)`.
This way a `Surface` won't need to hold a compositor instance but instead only the `SurfaceView` will need to get it, using the `of` method indeed.
It would also be possible to detect if the app is running with the compositor or not because of `Compositor.maybeOf(context)` returning null if no compositor is available for example

**Comment by HrX03** [comment]
I wonder if it would be appropriate to introduce a proper event api in this pr or it would be better to have it as a different pr. If a different pr would be optimal then i could work on it once this pr gets eventually merged

File: compositor_dart/lib/compositor_dart.dart (hunk 2 above)
Comment (hansihe) [reply compositor_dart/lib/compositor_dart.dart:31]: I would prefer not using the widget tree to pass around the compositor instance.

The main reason for this is that the compositor instance inherently has another lifetime compared to something in the widget tree.

It creates a platform channel and uses this to communicate with native code on the other side for the lifetime of the isolate. It also accumulates and maintains state for its whole lifetime, something conceptually incompatible with being maintained as part of the widget tree. It is analogous to the various `Binding`s in the flutter framework. (`WidgetBinding` for instance)

It's just inherently a singleton with a global lifetime, and I think we would save ourselves some pain and complexity in the long run by just treating it as such from the beginning.

That being said, whether we should pass it around as a part of the surface objects or access it as a global every time is still an open question for me. Having written this out and thought through it a bit more I'm leaning towards the global route.

**Comment by HrX03** [comment]
I already did an event api for utopia and i think we could lean towards that kind of architecture. Basically, you add a listener of sorts to the event source which is nothing more than a callback method that gets called each time any event happens. It has one single params which is a subclass of an abstract `Event` class that represents the kind of event that was generated. You just have to do some type checking to see which event was generated, eventually cast the param (tho usually it's automatically done as long as it's locally scoped) and then access the instance params and do your thing.

Basically it would be something like `Compositor.addEventHandler(callback)` where callback is of type `void Function(Event)`

File: compositor_dart/lib/compositor_dart.dart (hunk 2 above)
Comment (HrX03) [reply compositor_dart/lib/compositor_dart.dart:31]: Fair point then, i'm more leaning towards the global approach myself too

File: src/surface.c (hunk 3 above)
Comment (charafau) [reply src/surface.c:42]: sorry for late response. when I leave code as it was I get: 
```
00:00:11.553 [../src/surface.c:74] viewhandle 2
simple: types/xdg_shell/wlr_xdg_toplevel.c:516: wlr_xdg_toplevel_set_activated: Assertion `surface->role == WLR_XDG_SURFACE_ROLE_TOPLEVEL' failed.
00:00:11.553 [../src/flutter_wlroots.c:196] DART [flutter] handled call surface_map
```

File: compositor_dart/lib/compositor_dart.dart [hunk 4]
```diff
@@ -20,16 +21,45 @@ class Surface {
...
+
+  @override
+  int get hashCode {
+    return handle.hashCode ^
+        pid.hashCode ^
+        gid.hashCode ^
+        uid.hashCode ^
+        isPopup.hashCode ^
+        parentHandle.hashCode;
+  }
```
Comment (hansihe) [inline compositor_dart/lib/compositor_dart.dart:57]: If everything else is functioning as it should, there should be one and only one Surface class instance for every handle on the native side.

It should be safe to compare Surface classes by identity, so we probably don't need these implementations.

**Comment by charafau** [comment]
There's some breakage after merge. Need to fix that
